src/librustc_trans/glue.rs

```diff
@@ -205,11 +210,10 @@ pub fn implement_drop_glue<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>, g: DropGlueKi
     };
 
     let bcx = match t.sty {
-        ty::TyBox(content_ty) => {
-            // Support for TyBox is built-in and its drop glue is
-            // special. It may move to library and have Drop impl. As
-            // a safe-guard, assert TyBox not used with TyContents.
-            assert!(!skip_dtor);
```

Comment:
Do you mean something like
```
--- a/src/librustc_trans/glue.rs
+++ b/src/librustc_trans/glue.rs
@@ -210,7 +210,7 @@ pub fn implement_drop_glue<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>, g: DropGlueKi
     };

     let bcx = match t.sty {
-        ty::TyAdt(def, _) if def.is_box() => {
+        ty::TyAdt(def, _) if def.is_box()  && !skip_dtor => {
             // Support for Box is built-in as yet and its drop glue is special
             // despite having a dummy Drop impl in the library.
             let content_ty = t.boxed_ty();
```
?

I.e. if we get here with `skip_dtor == false`, we generate full drop glue (including both dropping the inner type and deallocation), otherwise we generate "structural drop" only, which is noop for Box.